gamemodes/terrortown/gamemode/server/sv_player_ext.lua

```diff
@@ -789,119 +795,261 @@ local function FindCorpse(ply)
--- @realm server
-function plymeta:Revive(delay, fn, check, needcorpse, force, onFail)
+-- @param [default=3]number delay The delay of the revive
+-- @param [opt]function OnRevive The @{function} that should be run if the @{Player} revives
+-- @param [opt]function DoCheck An additional checking @{function}
+-- @param [default=false]boolean needsCorpse Whether the dead @{Player} @{CORPSE} is needed
+-- @param [default=false]boolean blockRound Stops the round from ending if this is set to true until the player is alive again
+-- @param [opt]function OnFail This @{function} is called if the revive fails
+-- @param [opt]Vector spawnPos The position where the player should be spawned, accounts for minor obstacles
+-- @param [opt]Angle spawnEyeAngle The eye angles of the revived players
+-- @realm server
+function plymeta:Revive(delay, OnRevive, DoCheck, needsCorpse, blockRound, OnFail, spawnPos, spawnEyeAngle)
 	local ply = self
 	local name = "TTT2RevivePlayer" .. ply:EntIndex()
 
-	if timer.Exists(name) or ply.reviving then return end
+	if ply:IsReviving() then return end
 
-	if force then
-		ply.forceRevive = true
-	end
+	delay = delay or 3
+
+	ply:SetReviving(true)
+	ply:SetBlockingRevival(blockRound)
+	ply:SetRevivalStartTime(CurTime())
+	ply:SetRevivalDuration(delay)
 
-	ply.reviving = true
+	ply.OnReviveFailedCallback = OnFail
 
 	timer.Create(name, delay, 1, function()
 		if not IsValid(ply) then return end
 
-		ply.forceRevive = nil
-		ply.reviving = nil
+		ply:SetReviving(false)
+		ply:SetBlockingRevival(false)
+		ply:SendRevivalReason(nil)
 
-		if not isfunction(check) or check(ply) then
-			local corpse = FindCorpse(ply)
+		ply.OnReviveFailedCallback = nil
 
-			if needcorpse and (not IsValid(corpse) or corpse:IsOnFire()) then
-				ply:ChatPrint("You have not been revived because your body no longer exists.")
+		if not isfunction(DoCheck) or DoCheck(ply) then
+			local corpse = FindCorpse(ply)
 
-				timer.Remove(name) -- TODO needed?
+			if needsCorpse and (not IsValid(corpse) or corpse:IsOnFire()) then
+				OnReviveFailed(ply, OnFail, "message_revival_failed_missing_body")
 
 				return
 			end
 
 			ply:SpawnForRound(true)
 
-			if IsValid(corpse) then
-				local spawnPos = FindCorpsePosition(corpse)
-				if spawnPos then
-					ply:SetPos(spawnPos)
-					ply:SetEyeAngles(Angle(0, corpse:GetAngles().y, 0))
-				end
+			if not spawnPos and IsValid(corpse) then
+				spawnPos = corpse:GetPos()
+				spawnEyeAngle = Angle(0, corpse:GetAngles().y, 0)
 			end
 
-			hook.Call("PlayerLoadout", GAMEMODE, ply)
+			spawnPos = spawnPos or ply:GetDeathPosition()
+			spawnPos = spawn.MakeSpawnPointSafe(self, spawnPos)
+
+			if not spawnPos then
+				local spawnEntity = spawn.GetRandomPlayerSpawnEntity(ply)
 
+				spawnPos = spawnEntity:GetPos()
+				spawnEyeAngle = spawnEntity:EyeAngles()
+			end
+
+			ply:SetPos(spawnPos)
+			ply:SetEyeAngles(spawnEyeAngle or Angle(0, 0, 0))
 			ply:SetMaxHealth(100)
 
+			hook.Run("PlayerLoadout", ply)
```

Comment:
Shouldn't this be handled automatically by `GM:PlayerSpawn`? Ref.: https://wiki.facepunch.com/gmod/GM:PlayerSpawn

Reply:
I have no idea, this was in your revive function and I though you put your thought in it. I can test if it works without.

Reply:
I can't see it in the old code, GitHub displays that you added it :o

Reply:
Because I moved it to another file and put it in seperate functions. But the overall spawn logic isn't new, it's just a bit refactored

Reply:
See: <https://github.com/TTT-2/TTT2/pull/528/files#diff-bb7d466d24e307b3292a4722e8ade165L874>